Implement missing ready states transition

pylbrecht opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #31415

<!-- Either: -->
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

pylbrecht committed:
Expect video_size_preserved_after_ended.html to pass

eerii commented:
I tested the changes locally and they seem to be going in a great direction. At least it fixes the problem I was having in https://github.com/servo/servo/issues/31416#issuecomment-1987275602 with the `onloadedmetadata` event firing before the metadata was loaded (and thus failing the size asserts)!

pylbrecht committed:
Add VideoTrackList::remove()

Implementation is still pretty crude (panics if `idx` is out of bounds).
It's good enough for prototyping though. I will improve it once I get
things working.

pylbrecht committed:
Add debug! logs

pylbrecht committed:
Don't change ready state in HTMLMediaElementFetchListener

pylbrecht committed:
Remove video track when playback ended

pylbrecht committed:
Add AudioTrackList::remove()

pylbrecht committed:
Remove audio track when playback ended

pylbrecht committed:
Fire removetrack event for video track

pylbrecht committed:
Fire removetrack for audio track

pylbrecht committed:
Don't unwrap selected_index()

Fixes CRASH /html/semantics/embedded-content/the-video-element/resize-during-playback.html

pylbrecht committed:
Add spec reference

pylbrecht committed:
Update expectations on tests using track.onload

pylbrecht committed:
Add debug!() to queue_simple_event()

To narrow down the source for TIMEOUTs, I want to know which events are
queued.

pylbrecht committed:
Advance ready state for PlaybackState::Playing

Strictly speaking for `PlaybackState::Playing` the ready state could be
any of

* `ReadyState::HaveCurrentData`
* `ReadyState::HaveFutureData`
* `ReadyState::HaveEnoughData`

if the previous ready state was at least `ReadyState::HaveMetadata`.

We already assume the best case (i.e. `ReadyState::HaveEnoughData`) for
`PlaybackState::Paused` and we are fixing another unexpected WPT test
result (namely `/css/selectors/media/media-playback-state.html`) by
doing the same for `PlaybackState::Playing`.

pylbrecht committed:
fixup! Update expectations on tests using track.onload

pylbrecht committed:
Signal EOS to player at the end of process_reponse_eof()

pylbrecht commented:
Fixed in https://github.com/servo/servo/pull/32643.
